src/extensions/ColorPalette3DExtensions/BaseExtension.js

```diff
@@ -12,6 +12,12 @@ const defaultProps = {
  * @property {Array<Array<number>>=} colors Array of colors to map channels to (RGB).
  * */
 const BaseExtension = class extends LayerExtension {
+  constructor(args) {
+    super(args);
+    // After deck.gl 8.8, it does not seem like this is always initialized.
+    this.opts = this.opts || {};
```

Comment:
Why not replicate the previous behavior exactly. It's more clear that `LayerExtension` only accepts a single argument, and it's `opts`.

```suggestion
  // After deck.gl 8.8, opts is not initialized if not provided
  constructor(opts = {}) {
    super(opts);
```

Reply:
While the parent constructor does not take [other arguments](https://github.com/visgl/deck.gl/blob/master/modules/core/src/lib/layer-extension.ts) I don't think I want to rely on this behavior being forever true and would rather not break things if eventually there is some default argument that `deck.gl` adds in.  Does that make sense or am I being overly cautious?

Reply:
The ideal way to deal with this would be with static types (and then if it changed a type-checker would let us know), so I don't think that any choice w/o that is future proof (i.e., who is to say `opts` won't be changed as a property). My suggestion was to make it explicit what behavior we are relying on, but I do see your point. In that case, I'd probably forward all arguments (instead of just the first):

```typescript
  constructor(...args) {
    super(...args);
    // After deck.gl 8.8, it does not seem like this is always initialized.
    this.opts = this.opts || {};
```

The other thing I notice is that we don't need to rely on `opts` at all? We could just require that classes that extend from `BaseExtension` implement a `rendering` property?

```typescript
const MinimumIntensityProjectionExtension = class extends BaseExtension {
  rendering = { _BEFORE_RENDER, _RENDER, _AFTER_RENDER };
};
```

Reply:
I think sticking with `opts` is best practice: https://deck.gl/docs/developer-guide/custom-layers/layer-extensions#constructor-1

Reply:
> I think sticking with `opts` is best practice

I don't understand where this is expressed in the documentation. What is the best practice in this context?

> "Note that if two extension instances are of the same class and have the same `opts` object, they are considered equal during layer updates."

Each class which derives from `BaseExtension` is a unique; that alone should be sufficient for deck.gl to recognize they are different. My reading of `opts` is that it should be used to provide options for the extension (i.e., something that an end user could pass in which causes the shader to re-compile). The `_BEFORE_RENDER`/`_RENDER`/`_AFTER_RENDER` properties are all static (defined as a part of the class definition) and not meant to be overwritten by an end user.

> "Creating a new extension instance with different options can potentially lead to the layers that use this extension to recompile their shaders, which may be an expensive operation. `opts` is generally only used for options that require shader updates."

Our shaders don't require any `opts` really, since all the rendering is statically defined.

Reply:
That makes sense, will update.